enhancements/storage/lso-recreate-symlinks.md

```diff
@@ -0,0 +1,156 @@
+
+### User Stories
+
+* As an administrator, I want the option to switch to more stable symlinks for LSO PV's prior to upgrading OCP to a version which may cause existing by-id symlinks to change.
+* As an administrator, I want a way to recover existing LSO PV's when existing by-id symlinks change unexpectedly.
+
+### Goals
+
+* Preventative opt-in to recreate symlinks for existing LSO PV's based on the by-id symlink recommended by LSO as the most stable in relative terms.
+* Recovery mechanism to recreate symlinks for existing LSO PV's after the by-id symlinks have already changed.
+* Avoid manual changes on the node or other one-off recovery workarounds.
+
+### Non-Goals
+
+* LSO will not format or partition the device to maintain its own on-disk identifier outside of consumer control because this fundamentally alters the contract LSO offers to consumers.
+* LSO will not automatically recreate symlinks without explicit administrator opt-in because there may be other valid ways to resolve the issue and LSO does not have the context to make the right call in all cases.
+* LSO will not (at this time) offer a more fine-grained filtering mechanism to allow administrators to influence which symlinks should be preferred by LSO.
+
+## Proposal
+
+### Workflow Description
+
+LSO's diskmaker will detect all valid by-id symlinks for each PV and add them as an annotation on the PV. It will also annotate the by-id symlink currently in use. For filesystem volumes, it will annotate the UUID of the filesystem to help find the corresponding disk. This means each PV will have up to three new annotations added by diskmaker:
+
+* `storage.openshift.com/current-link-target`: current by-id symlink in use for the device
+* `storage.openshift.com/dev-disk-by-id-list`: list of valid by-id symlinks for the device
+* `storage.openshift.com/dev-disk-by-uuid`: by-uuid symlink corresponding to the device (when applicable)
+
+LSO will throw an alert if `current-link-target` does not match the recommended symlink in `dev-disk-by-id-list`. `dev-disk-by-uuid` is not available for raw block volumes, but may still be used for filesystem volumes to ensure detected symlinks match the device with the correct on-disk identifier.
+
+In response to the alert, the administrator can review the annotations on the PV, and then add a `storage.openshift.com/recreate-symlink` annotation to the PV to tell diskmaker to recreate the symlink. This can be done to proactively switch to the recommended symlink, or it can be done reactively when a symlink is no longer valid (assuming there is another known valid by-id symlink). Diskmaker will recreate the symlink pointing to the new by-id symlink and remove the `recreate-symlink` annotation when it is complete.
+
```

Comment:
I know we talked about this, But I guess in future if we wanted to give control over symlink that needs to be picked to customers, we could define another annotation which specifies user specified policy?  I know, one of the issues we discussed if we tried to do is, user specified preferred policy will be lost when PV is re-created but may be we could solve this by persisting `{symlinkBaseName: <preferred_policy>}` mapping to a json file.

We are hoping that what we are doing currently will be *enough* to steer us away from most of the issues.